README.md
<!-- Please use absolute URLs for all links so that they work when the content of this page is rendered in docs.pact.io -->

# Pact Go

Golang version of [Pact](http://pact.io). Pact is a contract testing framework for HTTP APIs and non-HTTP asynchronous messaging systems.

Enables consumer driven contract testing, providing a mock service and
DSL for the consumer project, and interaction playback and verification for the service Provider project.

[![Build Status](https://travis-ci.org/ray-xu-deltatre/pact-go.svg?branch=master)](https://travis-ci.org/ray-xu-deltatre/pact-go)
[![Build status](https://ci.appveyor.com/api/projects/status/lg02mfcmvr3e8w5n?svg=true)](https://ci.appveyor.com/project/mefellows/pact-go)
[![Coverage Status](https://coveralls.io/repos/github/ray-xu-deltatre/pact-go/badge.svg?branch=HEAD)](https://coveralls.io/github/ray-xu-deltatre/pact-go?branch=HEAD)
[![Go Report Card](https://goreportcard.com/badge/github.com/ray-xu-deltatre/pact-go)](https://goreportcard.com/report/github.com/ray-xu-deltatre/pact-go)
[![GoDoc](https://godoc.org/github.com/ray-xu-deltatre/pact-go?status.svg)](https://godoc.org/github.com/ray-xu-deltatre/pact-go)
[![slack](https://slack.pact.io/badge.svg)](https://slack.pact.io)
[![Known Vulnerabilities](https://snyk.io/test/github/ray-xu-deltatre/pact-go/badge.svg?targetFile=Gopkg.lock)](https://snyk.io/test/github/ray-xu-deltatre/pact-go?targetFile=Gopkg.lock)

## Introduction

From the [Pact website]:

> The Pact family of frameworks provide support for [Consumer Driven Contracts](http://martinfowler.com/articles/consumerDrivenContracts.html) testing.

> A Contract is a collection of agreements between a client (Consumer) and an API (Provider) that describes the interactions that can take place between them.

> Consumer Driven Contracts is a pattern that drives the development of the Provider from its Consumers point of view.

> Pact is a testing tool that guarantees those Contracts are satisfied.

Read [Getting started with Pact] for more information for beginners.

<p align="center">
  <a href="https://asciinema.org/a/180671">
    <img width="880" src="https://raw.githubusercontent.com/ray-xu-deltatre/pact-go/master/.github/pact-small.svg?sanitize=true"></img>
  </a>
</p>

## Table of Contents

<!-- TOC -->

- [Pact Go](#pact-go)
  - [Introduction](#introduction)
  - [Table of Contents](#table-of-contents)
  - [Versions](#versions)
  - [Installation](#installation)
    - [Go get](#go-get)
    - [Installation on \*nix](#installation-on-\nix)
  - [Using Pact](#using-pact)
  - [HTTP API Testing](#http-api-testing)
    - [Consumer Side Testing](#consumer-side-testing)
    - [Provider API Testing](#provider-api-testing)
      - [Provider Verification](#provider-verification)
      - [Provider States](#provider-states)
      - [Before and After Hooks](#before-and-after-hooks)
      - [Request Filtering](#request-filtering)
        - [Example: API with Authorization](#example-api-with-authorization)
      - [Pending Pacts](#pending-pacts)
      - [WIP Pacts](#wip-pacts)
      - [Lifecycle of a provider verification](#lifecycle-of-a-provider-verification)
    - [Publishing pacts to a Pact Broker and Tagging Pacts](#publishing-pacts-to-a-pact-broker-and-tagging-pacts)
      - [Publishing from Go code](#publishing-from-go-code)
      - [Publishing Provider Verification Results to a Pact Broker](#publishing-provider-verification-results-to-a-pact-broker)
      - [Publishing from the CLI](#publishing-from-the-cli)
      - [Using the Pact Broker with Basic authentication](#using-the-pact-broker-with-basic-authentication)
      - [Using the Pact Broker with Bearer Token authentication](#using-the-pact-broker-with-bearer-token-authentication)
  - [Asynchronous API Testing](#asynchronous-api-testing)
    - [Consumer](#consumer)
    - [Provider (Producer)](#provider-producer)
    - [Pact Broker Integration](#pact-broker-integration)
  - [Matching](#matching)
    - [Matching on types](#matching-on-types)
    - [Matching on arrays](#matching-on-arrays)
    - [Matching by regular expression](#matching-by-regular-expression)
    - [Match common formats](#match-common-formats)
      - [Auto-generate matchers from struct tags](#auto-generate-matchers-from-struct-tags)
  - [Tutorial (60 minutes)](#tutorial-60-minutes)
  - [Examples](#examples)
    - [HTTP APIs](#http-apis)
    - [Asynchronous APIs](#asynchronous-apis)
    - [Integrated examples](#integrated-examples)
  - [Troubleshooting](#troubleshooting)
      - [Splitting tests across multiple files](#splitting-tests-across-multiple-files)
      - [Output Logging](#output-logging)
      - [Check if the CLI tools are up to date](#check-if-the-cli-tools-are-up-to-date)
      - [Disable CLI checks to speed up tests](#disable-cli-checks-to-speed-up-tests)
      - [Re-run a specific provider verification test](#re-run-a-specific-provider-verification-test)
    - [Verifying APIs with a self-signed certificate](#verifying-apis-with-a-self-signed-certificate)
    - [Testing AWS API Gateway APIs](#testing-aws-api-gateway-apis)
  - [Contact](#contact)
  - [Documentation](#documentation)
  - [Roadmap](#roadmap)
  - [Contributing](#contributing)

<!-- /TOC -->
## Versions

<details><summary>Specification Compatibility</summary>

| Version | Stable | [Spec] Compatibility | Install            |
|---------|--------|----------------------|--------------------|
| 1.0.x   | Yes    | 2, 3\*               | See [installation] |
| 0.x.x   | Yes    | Up to v2             | 0.x.x [stable]     |

_\*_ v3 support is limited to the subset of functionality required to enable language inter-operable [Message support].

</details>

## Installation

1.  Download the latest [CLI tools] of the standalone tools and ensure the binaries are on your `PATH`:
1.  Unzip the package into a known location, and ensuring the `pact` and other binaries in the `bin` directory are on the `PATH`.
1.  Run `go get github.com/ray-xu-deltatre/pact-go@v1` to install the source packages

See below for how to automate this:

### Go get

Since `1.x.x` Pact is go-gettable, and uses tags for versioning, so `dep ensure --add github.com/ray-xu-deltatre/pact-go@1.0.0` or `go get gopkg.in/ray-xu-deltatre/pact-go.v1` is now possible.

See the [Changelog] for versions to pin to and their history.

### Installation on \*nix

The following will install pact binaries into `/opt/pact/bin`:

```sh
cd /opt
curl -fsSL https://raw.githubusercontent.com/pact-foundation/pact-ruby-standalone/master/install.sh | bash
export PATH=$PATH:/opt/pact/bin
go get github.com/ray-xu-deltatre/pact-go@v1
```

Test the installation:

```sh
pact help
```

## Using Pact

Pact supports [synchronous request-response style HTTP interactions](#http-api-testing) and has experimental support for [asynchronous interactions](#asynchronous-api-testing) with JSON-formatted payloads.

Pact Go runs as part of your regular Go tests.

## HTTP API Testing

### Consumer Side Testing

We'll run through a simple example to get an understanding the concepts:

1.  `go get github.com/ray-xu-deltatre/pact-go`
1.  `cd $GOPATH/src/github.com/ray-xu-deltatre/pact-go/examples/`
1.  `go test -v -run TestConsumer`.

The simple example looks like this:

```go
func TestConsumer(t *testing.T) {
	type User struct {
		Name     string `json:"name" pact:"example=billy"`
		LastName string `json:"lastName" pact:"example=sampson"`
	}

	// Create Pact connecting to local Daemon
	pact := &dsl.Pact{
		Consumer: "MyConsumer",
		Provider: "MyProvider",
		Host:     "localhost",
	}
	defer pact.Teardown()

	// Pass in test case. This is the component that makes the external HTTP call
	var test = func() (err error) {
		u := fmt.Sprintf("http://localhost:%d/foobar", pact.Server.Port)
		req, err := http.NewRequest("GET", u, strings.NewReader(`{"name":"billy"}`))
		if err != nil {
			return err
		}

		// NOTE: by default, request bodies are expected to be sent with a Content-Type
		// of application/json. If you don't explicitly set the content-type, you
		// will get a mismatch during Verification.
		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("Authorization", "Bearer 1234")

		if _, err = http.DefaultClient.Do(req); err != nil {
			return err
		}
	}

	// Set up our expected interactions.
	pact.
		AddInteraction().
		Given("User foo exists").
		UponReceiving("A request to get foo").
		WithRequest(dsl.Request{
			Method:  "GET",
			Path:    dsl.String("/foobar"),
			Headers: dsl.MapMatcher{"Content-Type": dsl.String("application/json"), "Authorization": dsl.String("Bearer 1234")},
			Body: map[string]string{
				"name": "billy",
			},
		}).
		WillRespondWith(dsl.Response{
			Status:  200,
			Headers: dsl.MapMatcher{"Content-Type": dsl.String("application/json")},
			Body:    dsl.Match(&User{}),
		})

	// Run the test, verify it did what we expected and capture the contract
	if err := pact.Verify(test); err != nil {
		log.Fatalf("Error on Verify: %v", err)
	}

	return nil
}

```

### Provider API Testing

1.  `go get github.com/ray-xu-deltatre/pact-go`
1.  `cd $GOPATH/src/github.com/ray-xu-deltatre/pact-go/examples/`
1.  `go test -v -run TestProvider`.

Here is the Provider test process broker down:

1.  Start your Provider API:

    You need to be able to first start your API in the background as part of your tests
    before you can run the verification process. Here we create `startServer` which can be
    started in its own goroutine:

    ```go
    var lastName = "" // User doesn't exist
    func startServer() {
      mux := http.NewServeMux()

      mux.HandleFunc("/users", func(w http.ResponseWriter, req *http.Request) {
        w.Header().Add("Content-Type", "application/json")
        fmt.Fprintf(w, fmt.Sprintf(`{"lastName":"%s"}`, lastName))
      })
      log.Fatal(http.ListenAndServe(":8000", mux))
    }
    ```

2)  Verify provider API

    You can now tell Pact to read in your Pact files and verify that your API will
    satisfy the requirements of each of your known consumers:

    ```go
    func TestProvider(t *testing.T) {

      // Create Pact connecting to local Daemon
      pact := &dsl.Pact{
        Provider: "MyProvider",
      }

      // Start provider API in the background
      go startServer()

      // Verify the Provider using the locally saved Pact Files
      pact.VerifyProvider(t, types.VerifyRequest{
        ProviderBaseURL:        "http://localhost:8000",
        PactURLs:               []string{filepath.ToSlash(fmt.Sprintf("%s/myconsumer-myprovider.json", pactDir))},
        StateHandlers: types.StateHandlers{
          // Setup any state required by the test
          // in this case, we ensure there is a "user" in the system
          "User foo exists": func() error {
            lastName = "crickets"
            return nil
          },
        },
      })
    }
    ```

The `VerifyProvider` will handle all verifications, treating them as subtests
and giving you granular test reporting. If you don't like this behaviour, you may call `VerifyProviderRaw` directly and handle the errors manually.

Note that `PactURLs` may be a list of local pact files or remote based
urls (e.g. from a
[Pact Broker](http://docs.pact.io/documentation/sharings_pacts.html)).

#### Provider Verification

When validating a Provider, you have 3 options to provide the Pact files:

1.  Use `PactURLs` to specify the exact set of pacts to be replayed:

    ```go
    pact.VerifyProvider(t, types.VerifyRequest{
    	ProviderBaseURL:        "http://myproviderhost",
    	PactURLs:               []string{"http://broker/pacts/provider/them/consumer/me/latest/dev"},
    	BrokerUsername:         os.Getenv("PACT_BROKER_USERNAME"),
    	BrokerPassword:         os.Getenv("PACT_BROKER_PASSWORD"),
    	BrokerToken:            os.Getenv("PACT_BROKER_TOKEN"),
    })
    ```

1.  Use `BrokerURL` to automatically find all of the latest consumers:

    ```go
    pact.VerifyProvider(t, types.VerifyRequest{
    	ProviderBaseURL:        "http://myproviderhost",
    	BrokerURL:              "http://brokerHost",
    	BrokerUsername:         os.Getenv("PACT_BROKER_USERNAME"),
    	BrokerPassword:         os.Getenv("PACT_BROKER_PASSWORD"),
    	BrokerToken:            os.Getenv("PACT_BROKER_TOKEN"),
    })
    ```

1.  Use `BrokerURL` and `Tags` to automatically find all of the latest consumers given one or more tags:

    ```go
    pact.VerifyProvider(t, types.VerifyRequest{
    	ProviderBaseURL:        "http://myproviderhost",
    	BrokerURL:              "http://brokerHost",
    	Tags:                   []string{"master", "prod"},
    	BrokerUsername:         os.Getenv("PACT_BROKER_USERNAME"),
    	BrokerPassword:         os.Getenv("PACT_BROKER_PASSWORD"),
    	BrokerToken:            os.Getenv("PACT_BROKER_TOKEN"),
    })
    ```

Options 2 and 3 are particularly useful when you want to validate that your
Provider is able to meet the contracts of what's in Production and also the latest
in development.

See this [article](http://rea.tech/enter-the-pact-matrix-or-how-to-decouple-the-release-cycles-of-your-microservices/)
for more on this strategy.

#### Provider States

If you have defined any states (as denoted by a `Given()`) in your consumer tests, the `Verifier` can put the provider into the correct state prior to sending the actual request for validation. For example, the provider can use the state to mock away certain database queries. To support this, set up a `StateHandler` for each state using hooks on the `StateHandlers` property. Here is an example:

```go
pact.VerifyProvider(t, types.VerifyRequest{
	...
  StateHandlers: types.StateHandlers{
		"User jmarie exists": func() error {
			userRepository = jmarieExists
			return nil
		},
		"User jmarie is unauthenticated": func() error {
			userRepository = jmarieUnauthorized
			token = "invalid"

			return nil
		},
		"User jmarie does not exist": func() error {
			userRepository = jmarieDoesNotExist
			return nil
		},
		...
	},
})
```

As you can see, for each state (`"User jmarie exists"` etc.) we configure the local datastore differently. If this option is not configured, the `Verifier` will ignore the provider states defined in the pact and log a warning.

Note that if the State Handler errors, the test will exit early with a failure.

Read more about [Provider States](https://docs.pact.io/getting_started/provider_states).

#### Before and After Hooks

Sometimes, it's useful to be able to do things before or after a test has run, such as reset a database, log a metric etc. A `BeforeEach` runs before any other part of the Pact test lifecycle, and a `AfterEach` runs as the last step before returning the verification result back to the test.

You can add them to your Verification as follows:

```go
	pact.VerifyProvider(t, types.VerifyRequest{
		...
		BeforeEach: func() error {
			fmt.Println("before hook, do something")
			return nil
		},
		AfterEach: func() error {
			fmt.Println("after hook, do something")
			return nil
		},
	})
```

If the Hook errors, the test will fail.

#### Request Filtering

Sometimes you may need to add things to the requests that can't be persisted in a pact file. Examples of these are authentication tokens with a small life span. e.g. an OAuth bearer token: `Authorization: Bearer 0b79bab50daca910b000d4f1a2b675d604257e42`.

For these cases, we have two facilities that should be carefully used during verification:

1. the ability to specify custom headers to be sent during provider verification. The flag to achieve this is `CustomProviderHeaders`.
1. the ability to modify a request/response and change the payload. The parameter to achieve this is `RequestFilter`.

Read on for more.

##### Example: API with Authorization

**Custom Headers**:

This header will always be sent for each and every request, and can't be dynamic. For example:

```go
  pact.VerifyProvider(t, types.VerifyRequest{
    ...
    CustomProviderHeaders:  []string{"Authorization: Bearer 0b79bab50daca910b000d4f1a2b675d604257e42"},
  })
```

As you can see, this is your opportunity to modify\add to headers being sent to the Provider API, for example to create a valid time-bound token.

**Request Filters**

_WARNING_: This should only be attempted once you know what you're doing!

Request filters are custom middleware, that are executed for each request, allowing `token` to change between invocations. Request filters can change the request coming in, _and_ the response back to the verifier. It is common to pair this with `StateHandlers` as per above, that can set/expire the token
for different test cases:

```go
  pact.VerifyProvider(t, types.VerifyRequest{
    ...
    RequestFilter: func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			r.Header.Add("Authorization", fmt.Sprintf("Bearer %s", token))
			next.ServeHTTP(w, r)
		})
	}
  })
```

_Important Note_: You should only use this feature for things that can not be persisted in the pact file. By modifying the request, you are potentially modifying the contract from the consumer tests!

#### Pending Pacts
_NOTE_: This feature is currently only available on [Pactflow]

Pending pacts is a feature that allows consumers to publish new contracts or changes to existing contracts without breaking Provider's builds. It does so by flagging the contract as "unverified" in the Pact Broker the first time a contract is published. A Provider can then enable a behaviour (via `EnablePending: true`) that will still perform a verification (and thus share the results back to the broker) but _not_ fail the verification step itself.

This enables safe introduction of new contracts into the system, without breaking Provider builds, whilst still providing feedback to Consumers as per before.

See the [docs](https://docs.pact.io/pending) and this [article](http://blog.pact.io/2020/02/24/how-we-have-fixed-the-biggest-problem-with-the-pact-workflow/) for more background.

#### WIP Pacts

_NOTE_: This feature is currently only available on [Pactflow]

WIP Pacts builds upon pending pacts, enabling provider tests to pull in _any_ contracts applicable to the provider regardless of the `tag` it was given. This is useful, because often times consumers won't follow the exact same tagging convention and so their workflow would be interrupted. This feature enables any pacts determined to be "work in progress" to be verified by the Provider, without causing a build failure. You can enable this behaviour by specifying a valid `time.Time` field for `IncludeWIPPactsSince`. This sets the start window for which new WIP pacts will be pulled down for verification, regardless of the tag.

See the [docs](https://docs.pact.io/wip) and this [article](http://blog.pact.io/2020/02/24/introducing-wip-pacts/) for more background.

#### Lifecycle of a provider verification

For each _interaction_ in a pact file, the order of execution is as follows:


CHANGELOG.md
# Changelog
Do this to generate your change history

    git log --pretty=format:'  * [%h](https://github.com/ray-xu-deltatre/pact-go/commit/%h) - %s (%an, %ad)' vX.Y.Z..HEAD | egrep -v "wip(:|\()" | grep -v "docs(" | grep -v "chore(" | grep -v Merge | grep -v "test("

## Versions

### v1.5.2 (25 March 2021)
  * [bae7fd8](https://github.com/ray-xu-deltatre/pact-go/commit/bae7fd8) - chore: demonstrate use of loglevel and log dir for TLS provider (Matt Fellows, Thu Mar 25 13:19:45 2021 +1100)
  * [8ed3655](https://github.com/ray-xu-deltatre/pact-go/commit/8ed3655) - chore: add log level to verifier CLI (Matt Fellows, Thu Mar 25 13:15:44 2021 +1100)
  * [a461ee3](https://github.com/ray-xu-deltatre/pact-go/commit/a461ee3) - fix: pass PactLogDir to verifier binary. Fixes #156 (Matt Fellows, Thu Mar 25 13:14:45 2021 +1100)
  * [2354ed9](https://github.com/ray-xu-deltatre/pact-go/commit/2354ed9) - chore: add linter to CI (Matt Fellows, Fri Oct 30 08:58:52 2020 +1100)

### v1.5.1 (09 October 2020)
  * [65822cd](https://github.com/ray-xu-deltatre/pact-go/commit/65822cd) - fix: misleading error message in certain verification scenarios (Matt Fellows, Fri Oct 9 23:59:37 2020 +1100)

### v1.5.0 (08 October 2020)
  * [d3abbd8](https://github.com/ray-xu-deltatre/pact-go/commit/d3abbd8) - feat: allow setting of ProviderTags on VerifyMessageRequest (James Halsall, Tue Sep 15 08:57:24 2020 +0100)
  * [d3c15fb](https://github.com/ray-xu-deltatre/pact-go/commit/d3c15fb) - docs: close img tag so it is valid mdx for docs.pact.io (Beth Skurrie, Sat Aug 15 18:02:16 2020 +1000)

### v1.4.4 (14 August 2020)
  * [a0227bb](https://github.com/ray-xu-deltatre/pact-go/commit/a0227bb) - fix: make tests fail on non-zero status (Matt Fellows, Fri Aug 7 14:37:42 2020 +1000)
  * [a7d9596](https://github.com/ray-xu-deltatre/pact-go/commit/a7d9596) - fix: timeout in integration test (Matt Fellows, Fri Aug 7 14:18:24 2020 +1000)
  * [4fe810d](https://github.com/ray-xu-deltatre/pact-go/commit/4fe810d) - fix: race in v2 client (Matt Fellows, Sun Aug 2 16:51:43 2020 +1000)
  * [2ea979c](https://github.com/ray-xu-deltatre/pact-go/commit/2ea979c) - fix: increase max size of pact output buffer (Cody Tatman, Thu Aug 6 18:11:26 2020 -0700)
  * [e25aa73](https://github.com/ray-xu-deltatre/pact-go/commit/e25aa73) - chore: avoid recompiling struct tag matching regexes (Michał Klimuk, Wed Jul 29 10:50:14 2020 +0200)
  * [1c18257](https://github.com/ray-xu-deltatre/pact-go/commit/1c18257) - fix: address various misspellings (Ben Wells, Fri Jul 17 12:55:33 2020 +0100)
  * [c77c0fd](https://github.com/ray-xu-deltatre/pact-go/commit/c77c0fd) - chore: return error if timeout occurs during provider verification (Matt Fellows, Thu Jul 16 17:14:31 2020 +1000)
  * [68baf93](https://github.com/ray-xu-deltatre/pact-go/commit/68baf93) - chore: wrap snyk in guard to prevent PRs failing (Matt Fellows, Sun Jul 5 22:48:32 2020 +1000)
  * [146de2c](https://github.com/ray-xu-deltatre/pact-go/commit/146de2c) - fix: handle JSON tags correctly (Jonathan Nowak, Tue Jun 30 07:28:00 2020 +0200)
  * [aec1bf2](https://github.com/ray-xu-deltatre/pact-go/commit/aec1bf2) - build: test golang 1.11 -> 1.14 (Matt Fellows, Mon Jun 29 09:57:01 2020 +1000)
  * [d93c924](https://github.com/ray-xu-deltatre/pact-go/commit/d93c924) - chore: remove build-only deps (Matt Fellows, Mon Jun 29 09:55:55 2020 +1000)
  * [9d3b451](https://github.com/ray-xu-deltatre/pact-go/commit/9d3b451) - chore: add semantic status checks for PRs and commit messages (Matt Fellows, Mon Jun 29 09:49:03 2020 +1000)
  * [88d6be4](https://github.com/ray-xu-deltatre/pact-go/commit/88d6be4) - build: add snyk to build (Matt Fellows, Mon Jun 29 09:47:03 2020 +1000)
  * [2b0ad4c](https://github.com/ray-xu-deltatre/pact-go/commit/2b0ad4c) - chore: remove Gopkg lock files (Matt Fellows, Mon Jun 29 09:25:24 2020 +1000)
  * [18398ce](https://github.com/ray-xu-deltatre/pact-go/commit/18398ce) - feat: update to standalone 1.86.0 and support new verifier options: --log-level and --log-dir (Matt Fellows, Sat Jun 27 14:21:33 2020 +1000)
  * [26aac7a](https://github.com/ray-xu-deltatre/pact-go/commit/26aac7a) - chore: add workflow to trigger sync to docs.pact.io when any markdown files change (Beth Skurrie, Fri Jun 26 15:14:33 2020 +1000)
  * [264394a](https://github.com/ray-xu-deltatre/pact-go/commit/264394a) - docs: move troubleshooting from wiki to repo (Beth Skurrie, Fri Jun 26 14:50:55 2020 +1000)
  * [9f54764](https://github.com/ray-xu-deltatre/pact-go/commit/9f54764) - docs: update links (Beth Skurrie, Fri Jun 26 14:49:31 2020 +1000)
  * [2fdb281](https://github.com/ray-xu-deltatre/pact-go/commit/2fdb281) - docs: update urls (Beth Skurrie, Fri Jun 26 14:46:45 2020 +1000)
  * [1fd82fb](https://github.com/ray-xu-deltatre/pact-go/commit/1fd82fb) - docs: use https instead of https for slack image (Beth Skurrie, Fri Jun 26 14:42:29 2020 +1000)
  * [7b0175f](https://github.com/ray-xu-deltatre/pact-go/commit/7b0175f) - docs: update for rending in docs.pact.io (Beth Skurrie, Fri Jun 26 14:42:00 2020 +1000)

### v1.4.3 (09 June 2020)
  * [4710769](https://github.com/ray-xu-deltatre/pact-go/commit/4710769) - chore: allow response body to be traced (Matt Fellows, Tue Jun 9 10:46:04 2020 +1000)

### v1.4.2 (09 June 2020)
  * [41c3c52](https://github.com/ray-xu-deltatre/pact-go/commit/41c3c52) - fix: use stderr scanner on captureing stderr (Shin'ya UEOKA, Mon Jun 8 23:07:01 2020 +0000)

### v1.4.1 (22 May 2020)
  * [0dfbb4b](https://github.com/ray-xu-deltatre/pact-go/commit/0dfbb4b) - fix: remove deadlock for large provider tests (Matt Fellows, Thu May 21 22:06:06 2020 +1000)
  * [118452e](https://github.com/ray-xu-deltatre/pact-go/commit/118452e) - chore: fix typo for no pacts error (Dom Udall, Tue May 19 14:11:03 2020 +0100)

doc.go
/*
Pact Go enables consumer driven contract testing, providing a mock service and
DSL for the consumer project, and interaction playback and verification
for the service provider project.

Consumer Tests

Consumer side Pact testing is an isolated test that ensures a given component
is able to collaborate with another (remote) component. Pact will automatically
start a Mock server in the background that will act as the collaborators' test
double.

This implies that any interactions expected on the Mock server will be validated,
meaning a test will fail if all interactions were not completed, or if unexpected
interactions were found:

A typical consumer-side test would look something like this:

	func TestLogin(t *testing.T) {

		// Create Pact client
		pact := Pact{
			Consumer: "My Consumer",
			Provider: "My Provider",
		}
		// Shuts down Mock Service when done
		defer pact.Teardown()

		// Pass in your test case as a function to Verify()
		var test = func() error {
			_, err := http.Get("http://localhost:8000/")
			return err
		}

    // Set up our expected interactions.
    pact.
      AddInteraction().
      Given("User foo exists").
      UponReceiving("A request to get foo").
      WithRequest(dsl.Request{
        Method:  "GET",
        Path:    dsl.String("/foobar"),
        Headers: dsl.MapMatcher{"Content-Type": "application/json"},
      }).
      WillRespondWith(dsl.Response{
        Status:  200,
        Headers: dsl.MapMatcher{"Content-Type": "application/json"},
        Body:    dsl.Match(&Foo{})
      })

    // Verify
    if err := pact.Verify(test); err != nil {
      log.Fatalf("Error on Verify: %v", err)
    }
  }

If this test completed successfully, a Pact file should have been written to
./pacts/my_consumer-my_provider.json containing all of the interactions
expected to occur between the Consumer and Provider.

Matching

In addition to verbatim value matching, you have 3 useful matching functions
in the `dsl` package that can increase expressiveness and reduce brittle test
cases.

	Term(example, matcher)	tells Pact that the value should match using a given regular expression, using `example` in mock responses. `example` must be a string.
	Like(content)		tells Pact that the value itself is not important, as long as the element _type_ (valid JSON number, string, object etc.) itself matches.
	EachLike(content, min)	tells Pact that the value should be an array type, consisting of elements like those passed in. `min` must be >= 1. `content` may be a valid JSON value: e.g. strings, numbers and objects.

Here is a complex example that shows how all 3 terms can be used together:

  	body :=
		Like(map[string]interface{}{
			"response": map[string]interface{}{
				"name": Like("Billy"),
        "type": Term("admin", "admin|user|guest"),
        "items": EachLike("cat", 2)
			},
		})

This example will result in a response body from the mock server that looks like:

  {
    "response": {
      "name": "Billy",
      "type": "admin",
      "items": [
        "cat",
        "cat"
      ]
    }
  }

See the examples in the dsl package and the matcher tests
(https://github.com/ray-xu-deltatre/pact-go/blob/master/dsl/matcher_test.go)
for more matching examples.

NOTE: You will need to use valid Ruby regular expressions
(http://ruby-doc.org/core-2.1.5/Regexp.html) and double escape backslashes.

Read more about flexible matching (https://github.com/pact-foundation/pact-ruby/wiki/Regular-expressions-and-type-matching-with-Pact.

Provider Tests

Provider side Pact testing, involves verifying that the contract - the Pact file
- can be satisfied by the Provider.

A typical Provider side test would like something like:

	func TestProvider_PactContract(t *testing.T) {
	  // Create Pact
		pact := Pact{}
		go startMyAPI("http://localhost:8000")

		pact.VerifyProvider(types.VerifyRequest{
			ProviderBaseURL:        "http://localhost:8000",
			PactURLs:               []string{"./pacts/my_consumer-my_provider.json"},
			ProviderStatesSetupURL: "http://localhost:8000/setup",
		})
	}

The `VerifyProvider` will handle all verifications, treating them as subtests
and giving you granular test reporting. If you don't like this behaviour, you may
call `VerifyProviderRaw` directly and handle the errors manually.

Note that `PactURLs` may be a list of local pact files or remote based
urls (possibly from a Pact Broker
- http://docs.pact.io/documentation/sharings_pacts.html).

Pact reads the specified pact files (from remote or local sources) and replays
the interactions against a running Provider. If all of the interactions are met
we can say that both sides of the contract are satisfied and the test passes.

Provider Verification

When validating a Provider, you have 3 options to provide the Pact files:

1. Use "PactURLs" to specify the exact set of pacts to be replayed:

	response, err = pact.VerifyProvider(types.VerifyRequest{
		ProviderBaseURL:        "http://myproviderhost",
		PactURLs:               []string{"http://broker/pacts/provider/them/consumer/me/latest/dev"},
		ProviderStatesSetupURL: "http://myproviderhost/setup",
		BrokerUsername:         os.Getenv("PACT_BROKER_USERNAME"),
		BrokerPassword:         os.Getenv("PACT_BROKER_PASSWORD"),
		BrokerToken:            os.Getenv("PACT_BROKER_TOKEN"),
	})

	2. Use "PactBroker" to automatically find all of the latest consumers:

	response, err = pact.VerifyProvider(types.VerifyRequest{
		ProviderBaseURL:        "http://myproviderhost",
		BrokerURL:              brokerHost,
		ProviderStatesSetupURL: "http://myproviderhost/setup",
		BrokerUsername:         os.Getenv("PACT_BROKER_USERNAME"),
		BrokerPassword:         os.Getenv("PACT_BROKER_PASSWORD"),
		BrokerToken:            os.Getenv("PACT_BROKER_TOKEN"),
	})

	3. Use "PactBroker" and "Tags" to automatically find all of the latest consumers:

	response, err = pact.VerifyProvider(types.VerifyRequest{
		ProviderBaseURL:        "http://myproviderhost",
		BrokerURL:              brokerHost,
		Tags:                   []string{"master", "sit4"},
		ProviderStatesSetupURL: "http://myproviderhost/setup",
		BrokerUsername:         os.Getenv("PACT_BROKER_USERNAME"),
		BrokerPassword:         os.Getenv("PACT_BROKER_PASSWORD"),
		BrokerToken:            os.Getenv("PACT_BROKER_TOKEN"),
	})

Options 2 and 3 are particularly useful when you want to validate that your
Provider is able to meet the contracts of what's in Production and also the latest
in development.

See this [article](http://rea.tech/enter-the-pact-matrix-or-how-to-decouple-the-release-cycles-of-your-microservices/)
for more on this strategy.

Provider States

Each interaction in a pact should be verified in isolation, with no context
maintained from the previous interactions. So how do you test a request that
requires data to exist on the provider? Provider states are how you achieve
this using Pact.

Provider states also allow the consumer to make the same request with different
expected responses (e.g. different response codes, or the same resource with a
different subset of data).

States are configured on the consumer side when you issue a dsl.Given() clause
with a corresponding request/response pair.

Configuring the provider is a little more involved, and (currently) requires
running an API endpoint to configure any [provider states](http://docs.pact.io/documentation/provider_states.html) during the
verification process. The option you must provide to the dsl.VerifyRequest
is:

  ProviderStatesSetupURL: 	POST URL to set the provider state (see types.ProviderState)

An example route using the standard Go http package might look like this:

  // Handle a request from the verifier to configure a provider state (ProviderStatesSetupURL)
  mux.HandleFunc("/setup", func(w http.ResponseWriter, req *http.Request) {
    w.Header().Add("Content-Type", "application/json")

    // Retrieve the Provider State
    var state types.ProviderState

    body, _ := ioutil.ReadAll(req.Body)
    req.Body.Close()
    json.Unmarshal(body, &state)

    // Configure database for different states
    if state.State == "User A exists" {
      svc.userDatabase = aExists
    } else if state.State == "User A is unauthorized" {
      svc.userDatabase = aUnauthorized
    } else {
      svc.userDatabase = aDoesNotExist
    }
  })

See the examples or read more at http://docs.pact.io/documentation/provider_states.html.

Publishing Pacts to a Broker and Tagging Pacts

See the Pact Broker (http://docs.pact.io/documentation/sharings_pacts.html)
documentation for more details on the Broker and this article
(http://rea.tech/enter-the-pact-matrix-or-how-to-decouple-the-release-cycles-of-your-microservices/)
on how to make it work for you.

Publishing using Go code:

	pact.PublishPacts(types.PublishRequest{
		PactBroker:             "http://pactbroker:8000",
		PactURLs:               []string{"./pacts/my_consumer-my_provider.json"},
		ConsumerVersion:        "1.0.0",
		Tags:                   []string{"master", "dev"},
	})

Publishing from the CLI:

Use a cURL request like the following to PUT the pact to the right location,
specifying your consumer name, provider name and consumer version.

	curl -v -XPUT \-H "Content-Type: application/json" \
	-d@spec/pacts/a_consumer-a_provider.json \
	http://your-pact-broker/pacts/provider/A%20Provider/consumer/A%20Consumer/version/1.0.0

Using the Pact Broker with Basic authentication

The following flags are required to use basic authentication when
publishing or retrieving Pact files to/from a Pact Broker:

	BrokerUsername	uername for Pact Broker basic authentication
	BrokerPassword	password for Pact Broker basic authentication

Output Logging

Pact Go uses a simple log utility (logutils - https://github.com/hashicorp/logutils)
to filter log messages. The CLI already contains flags to manage this,
should you want to control log level in your tests, you can set it like so:

	pact := Pact{
	  ...
		LogLevel: "DEBUG", // One of DEBUG, INFO, ERROR, NONE
	}
*/
package main
